In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

Section 1: Exploration (Q1–5)

1. Load the dataset and display the first 10 rows.

In [2]:
df = pd.read_csv('imdb_top_1000.csv')
df.head(10)

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142 min,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,"28,341,469"
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175 min,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,"134,966,411"
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,"534,858,444"
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202 min,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,"57,300,000"
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96 min,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,"4,360,000"
5,https://m.media-amazon.com/images/M/MV5BNzA5ZD...,The Lord of the Rings: The Return of the King,2003,U,201 min,"Action, Adventure, Drama",8.9,Gandalf and Aragorn lead the World of Men agai...,94.0,Peter Jackson,Elijah Wood,Viggo Mortensen,Ian McKellen,Orlando Bloom,1642758,"377,845,905"
6,https://m.media-amazon.com/images/M/MV5BNGNhMD...,Pulp Fiction,1994,A,154 min,"Crime, Drama",8.9,"The lives of two mob hitmen, a boxer, a gangst...",94.0,Quentin Tarantino,John Travolta,Uma Thurman,Samuel L. Jackson,Bruce Willis,1826188,"107,928,762"
7,https://m.media-amazon.com/images/M/MV5BNDE4OT...,Schindler's List,1993,A,195 min,"Biography, Drama, History",8.9,"In German-occupied Poland during World War II,...",94.0,Steven Spielberg,Liam Neeson,Ralph Fiennes,Ben Kingsley,Caroline Goodall,1213505,"96,898,818"
8,https://m.media-amazon.com/images/M/MV5BMjAxMz...,Inception,2010,UA,148 min,"Action, Adventure, Sci-Fi",8.8,A thief who steals corporate secrets through t...,74.0,Christopher Nolan,Leonardo DiCaprio,Joseph Gordon-Levitt,Elliot Page,Ken Watanabe,2067042,"292,576,195"
9,https://m.media-amazon.com/images/M/MV5BMmEzNT...,Fight Club,1999,A,139 min,Drama,8.8,An insomniac office worker and a devil-may-car...,66.0,David Fincher,Brad Pitt,Edward Norton,Meat Loaf,Zach Grenier,1854740,"37,030,102"


2. Check the shape (rows, columns) of the dataset.

In [3]:
df.shape

(1000, 16)

3.List all column names and their data types.

In [4]:
df.columns

Index(['Poster_Link', 'Series_Title', 'Released_Year', 'Certificate',
       'Runtime', 'Genre', 'IMDB_Rating', 'Overview', 'Meta_score', 'Director',
       'Star1', 'Star2', 'Star3', 'Star4', 'No_of_Votes', 'Gross'],
      dtype='object')

In [5]:
df.dtypes

Poster_Link       object
Series_Title      object
Released_Year     object
Certificate       object
Runtime           object
Genre             object
IMDB_Rating      float64
Overview          object
Meta_score       float64
Director          object
Star1             object
Star2             object
Star3             object
Star4             object
No_of_Votes        int64
Gross             object
dtype: object

4. Check for missing values in each column.

In [6]:
df.isnull().sum()

Poster_Link        0
Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Overview           0
Meta_score       157
Director           0
Star1              0
Star2              0
Star3              0
Star4              0
No_of_Votes        0
Gross            169
dtype: int64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Poster_Link    1000 non-null   object 
 1   Series_Title   1000 non-null   object 
 2   Released_Year  1000 non-null   object 
 3   Certificate    899 non-null    object 
 4   Runtime        1000 non-null   object 
 5   Genre          1000 non-null   object 
 6   IMDB_Rating    1000 non-null   float64
 7   Overview       1000 non-null   object 
 8   Meta_score     843 non-null    float64
 9   Director       1000 non-null   object 
 10  Star1          1000 non-null   object 
 11  Star2          1000 non-null   object 
 12  Star3          1000 non-null   object 
 13  Star4          1000 non-null   object 
 14  No_of_Votes    1000 non-null   int64  
 15  Gross          831 non-null    object 
dtypes: float64(2), int64(1), object(13)
memory usage: 125.1+ KB


5. Get summary statistics (.describe()) for numeric columns.

In [8]:
df.describe() #for all columns including text/categorical ones, use df.describe(include='all')

,IMDB_Rating,Meta_score,No_of_Votes
count,1000.000000,843.000000,1.000000e+03
mean,7.949300,77.971530,2.736929e+05
std,0.275491,12.376099,3.273727e+05
min,7.600000,28.000000,2.508800e+04
25%,7.700000,70.000000,5.552625e+04
50%,7.900000,79.000000,1.385485e+05
75%,8.100000,87.000000,3.741612e+05
max,9.300000,100.000000,2.343110e+06


Section 2: Cleaning (Q6–10)

6. Find how many rows have a missing Revenue value.

In [9]:
df['Gross'].isnull().sum()

np.int64(169)

7. Fill missing Revenue values with the median revenue.

In [10]:
df['Gross'] = df['Gross'].str.replace(',', '').astype(float)

In [11]:
df['Gross'] = df['Gross'].fillna(df['Gross'].median())
df['Gross']

0       28341469.0
1      134966411.0
2      534858444.0
3       57300000.0
4        4360000.0
          ...     
995     23530892.0
996     23530892.0
997     30500000.0
998     23530892.0
999     23530892.0
Name: Gross, Length: 1000, dtype: float64

8. Drop rows where Rating is missing (if any).

In [12]:
df['IMDB_Rating'] = df['IMDB_Rating'].dropna()

9. Check for and remove duplicate rows.

In [13]:
df = df.drop_duplicates()

10. Convert the Runtime column to integer type if it's stored as text (e.g., "142 min" → 142).

In [14]:
df['Runtime'] = df['Runtime'].str.replace('min','').astype(int)

In [15]:
df['Runtime'].dtype

dtype('int64')

Section 3: Filtering (Q11–15)

11. Find all movies released after 2010.

In [16]:
df[df['Released_Year'] > '2010']['Series_Title']

18               Hamilton
19           Gisaengchung
20        Soorarai Pottru
21           Interstellar
33                  Joker
              ...        
903        A Star Is Born
907                 50/50
910             Moneyball
911    La piel que habito
966             Apollo 13
Name: Series_Title, Length: 226, dtype: object

12. Find all movies with a rating above 8.0.

In [17]:
df[df['IMDB_Rating'] > 8.0]['Series_Title']

0           The Shawshank Redemption
1                      The Godfather
2                    The Dark Knight
3             The Godfather: Part II
4                       12 Angry Men
                   ...              
317       La passion de Jeanne d'Arc
318                       The Circus
319    Sunrise: A Song of Two Humans
320                      The General
321     Das Cabinet des Dr. Caligari
Name: Series_Title, Length: 322, dtype: object

13. Find all movies directed by a specific director of your choice (e.g., Christopher Nolan).

In [18]:
df[df['Director'] == 'Christopher Nolan']['Series_Title']

2            The Dark Knight
8                  Inception
21              Interstellar
36              The Prestige
63     The Dark Knight Rises
69                   Memento
155            Batman Begins
573                  Dunkirk
Name: Series_Title, dtype: object

14. Find movies where the genre contains "Action" (hint: .str.contains()).

In [19]:
df.head()

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,28341469.0
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,134966411.0
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,534858444.0
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,57300000.0
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,4360000.0


In [20]:
df[df['Genre'].str.contains('Action')]

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,534858444.0
5,https://m.media-amazon.com/images/M/MV5BNzA5ZD...,The Lord of the Rings: The Return of the King,2003,U,201,"Action, Adventure, Drama",8.9,Gandalf and Aragorn lead the World of Men agai...,94.0,Peter Jackson,Elijah Wood,Viggo Mortensen,Ian McKellen,Orlando Bloom,1642758,377845905.0
8,https://m.media-amazon.com/images/M/MV5BMjAxMz...,Inception,2010,UA,148,"Action, Adventure, Sci-Fi",8.8,A thief who steals corporate secrets through t...,74.0,Christopher Nolan,Leonardo DiCaprio,Joseph Gordon-Levitt,Elliot Page,Ken Watanabe,2067042,292576195.0
10,https://m.media-amazon.com/images/M/MV5BN2EyZj...,The Lord of the Rings: The Fellowship of the Ring,2001,U,178,"Action, Adventure, Drama",8.8,A meek Hobbit from the Shire and eight compani...,92.0,Peter Jackson,Elijah Wood,Ian McKellen,Orlando Bloom,Sean Bean,1661481,315544750.0
13,https://m.media-amazon.com/images/M/MV5BZGMxZT...,The Lord of the Rings: The Two Towers,2002,UA,179,"Action, Adventure, Drama",8.7,While Frodo and Sam edge closer to Mordor with...,87.0,Peter Jackson,Elijah Wood,Ian McKellen,Viggo Mortensen,Orlando Bloom,1485555,342551365.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
968,https://m.media-amazon.com/images/M/MV5BYjcxMz...,Falling Down,1993,R,113,"Action, Crime, Drama",7.6,An ordinary man frustrated with the various fl...,56.0,Joel Schumacher,Michael Douglas,Robert Duvall,Barbara Hershey,Rachel Ticotin,171640,40903593.0
979,https://m.media-amazon.com/images/M/MV5BZTllNW...,Lethal Weapon,1987,A,109,"Action, Crime, Thriller",7.6,Two newly paired cops who are complete opposit...,68.0,Richard Donner,Mel Gibson,Danny Glover,Gary Busey,Mitchell Ryan,236894,65207127.0
982,https://m.media-amazon.com/images/M/MV5BN2VlNj...,Mad Max 2,1981,A,96,"Action, Adventure, Sci-Fi",7.6,"In the post-apocalyptic Australian wasteland, ...",77.0,George Miller,Mel Gibson,Bruce Spence,Michael Preston,Max Phipps,166588,12465371.0
983,https://m.media-amazon.com/images/M/MV5BYTU2MW...,The Warriors,1979,UA,92,"Action, Crime, Thriller",7.6,"In the near future, a charismatic leader summo...",65.0,Walter Hill,Michael Beck,James Remar,Dorsey Wright,Brian Tyler,93878,22490039.0


15. Find movies with runtime between 90 and 120 minutes.

In [21]:
df[(df['Runtime'] >= 90) & (df['Runtime'] <= 120)]

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,4360000.0
26,https://m.media-amazon.com/images/M/MV5BYmJmM2...,La vita è bella,1997,U,116,"Comedy, Drama, Romance",8.6,When an open-minded Jewish librarian and his s...,59.0,Roberto Benigni,Roberto Benigni,Nicoletta Braschi,Giorgio Cantarini,Giustino Durano,623629,57598247.0
28,https://m.media-amazon.com/images/M/MV5BNjNhZT...,The Silence of the Lambs,1991,A,118,"Crime, Drama, Thriller",8.6,A young F.B.I. cadet must receive the help of ...,85.0,Jonathan Demme,Jodie Foster,Anthony Hopkins,Lawrence A. Bonney,Kasi Lemmons,1270197,130742922.0
34,https://m.media-amazon.com/images/M/MV5BOTA5ND...,Whiplash,2014,A,106,"Drama, Music",8.5,A promising young drummer enrolls at a cut-thr...,88.0,Damien Chazelle,Miles Teller,J.K. Simmons,Melissa Benoist,Paul Reiser,717585,13092000.0
35,https://m.media-amazon.com/images/M/MV5BMTYxND...,The Intouchables,2011,UA,112,"Biography, Comedy, Drama",8.5,After he becomes a quadriplegic from a paragli...,57.0,Olivier Nakache,Éric Toledano,François Cluzet,Omar Sy,Anne Le Ny,760360,13182281.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
989,https://m.media-amazon.com/images/M/MV5BYzZhOD...,The Long Goodbye,1973,A,112,"Comedy, Crime, Drama",7.6,Private investigator Philip Marlowe helps a fr...,87.0,Robert Altman,Elliott Gould,Nina van Pallandt,Sterling Hayden,Mark Rydell,26337,959000.0
993,https://m.media-amazon.com/images/M/MV5BYTE4YW...,Blowup,1966,A,111,"Drama, Mystery, Thriller",7.6,A fashion photographer unknowingly captures a ...,82.0,Michelangelo Antonioni,David Hemmings,Vanessa Redgrave,Sarah Miles,John Castle,56513,23530892.0
995,https://m.media-amazon.com/images/M/MV5BNGEwMT...,Breakfast at Tiffany's,1961,A,115,"Comedy, Drama, Romance",7.6,A young New York socialite becomes interested ...,76.0,Blake Edwards,Audrey Hepburn,George Peppard,Patricia Neal,Buddy Ebsen,166544,23530892.0
997,https://m.media-amazon.com/images/M/MV5BM2U3Yz...,From Here to Eternity,1953,Passed,118,"Drama, Romance, War",7.6,"In Hawaii in 1941, a private is cruelly punish...",85.0,Fred Zinnemann,Burt Lancaster,Montgomery Clift,Deborah Kerr,Donna Reed,43374,30500000.0


Section 4: Sorting & Selection (Q16–18)

16. Get the top 10 highest-rated movies.

In [22]:
df.sort_values('IMDB_Rating', ascending=False).head(9)

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,28341469.0
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,134966411.0
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,534858444.0
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,57300000.0
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,4360000.0
5,https://m.media-amazon.com/images/M/MV5BNzA5ZD...,The Lord of the Rings: The Return of the King,2003,U,201,"Action, Adventure, Drama",8.9,Gandalf and Aragorn lead the World of Men agai...,94.0,Peter Jackson,Elijah Wood,Viggo Mortensen,Ian McKellen,Orlando Bloom,1642758,377845905.0
6,https://m.media-amazon.com/images/M/MV5BNGNhMD...,Pulp Fiction,1994,A,154,"Crime, Drama",8.9,"The lives of two mob hitmen, a boxer, a gangst...",94.0,Quentin Tarantino,John Travolta,Uma Thurman,Samuel L. Jackson,Bruce Willis,1826188,107928762.0
7,https://m.media-amazon.com/images/M/MV5BNDE4OT...,Schindler's List,1993,A,195,"Biography, Drama, History",8.9,"In German-occupied Poland during World War II,...",94.0,Steven Spielberg,Liam Neeson,Ralph Fiennes,Ben Kingsley,Caroline Goodall,1213505,96898818.0
10,https://m.media-amazon.com/images/M/MV5BN2EyZj...,The Lord of the Rings: The Fellowship of the Ring,2001,U,178,"Action, Adventure, Drama",8.8,A meek Hobbit from the Shire and eight compani...,92.0,Peter Jackson,Elijah Wood,Ian McKellen,Orlando Bloom,Sean Bean,1661481,315544750.0


17. Get the top 10 movies by revenue.

In [23]:
df.sort_values('Gross',ascending=False).head(9)

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
477,https://m.media-amazon.com/images/M/MV5BOTAzOD...,Star Wars: Episode VII - The Force Awakens,2015,U,138,"Action, Adventure, Sci-Fi",7.9,"As a new threat to the galaxy rises, Rey, a de...",80.0,J.J. Abrams,Daisy Ridley,John Boyega,Oscar Isaac,Domhnall Gleeson,860823,936662225.0
59,https://m.media-amazon.com/images/M/MV5BMTc5MD...,Avengers: Endgame,2019,UA,181,"Action, Adventure, Drama",8.4,After the devastating events of Avengers: Infi...,78.0,Anthony Russo,Joe Russo,Robert Downey Jr.,Chris Evans,Mark Ruffalo,809955,858373000.0
623,https://m.media-amazon.com/images/M/MV5BMTYwOT...,Avatar,2009,UA,162,"Action, Adventure, Fantasy",7.8,A paraplegic Marine dispatched to the moon Pan...,83.0,James Cameron,Sam Worthington,Zoe Saldana,Sigourney Weaver,Michelle Rodriguez,1118998,760507625.0
60,https://m.media-amazon.com/images/M/MV5BMjMxNj...,Avengers: Infinity War,2018,UA,149,"Action, Adventure, Sci-Fi",8.4,The Avengers and their allies must be willing ...,68.0,Anthony Russo,Joe Russo,Robert Downey Jr.,Chris Hemsworth,Mark Ruffalo,834477,678815482.0
652,https://m.media-amazon.com/images/M/MV5BMDdmZG...,Titanic,1997,UA,194,"Drama, Romance",7.8,A seventeen-year-old aristocrat falls in love ...,75.0,James Cameron,Leonardo DiCaprio,Kate Winslet,Billy Zane,Kathy Bates,1046089,659325379.0
357,https://m.media-amazon.com/images/M/MV5BNDYxNj...,The Avengers,2012,UA,143,"Action, Adventure, Sci-Fi",8.0,Earth's mightiest heroes must come together an...,69.0,Joss Whedon,Robert Downey Jr.,Chris Evans,Scarlett Johansson,Jeremy Renner,1260806,623279547.0
891,https://m.media-amazon.com/images/M/MV5BMTEzNz...,Incredibles 2,2018,UA,118,"Animation, Action, Adventure",7.6,The Incredibles family takes on a new mission ...,80.0,Brad Bird,Craig T. Nelson,Holly Hunter,Sarah Vowell,Huck Milner,250057,608581744.0
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,534858444.0
582,https://m.media-amazon.com/images/M/MV5BMjEwMz...,Rogue One,2016,UA,133,"Action, Adventure, Sci-Fi",7.8,The daughter of an Imperial scientist joins th...,65.0,Gareth Edwards,Felicity Jones,Diego Luna,Alan Tudyk,Donnie Yen,556608,532177324.0


18. Select only Title, Year, and Rating columns and display the top 5 by rating.

In [24]:
df.sort_values('IMDB_Rating',ascending=False).head(4)[['Series_Title','Released_Year','IMDB_Rating']]

,Series_Title,Released_Year,IMDB_Rating
0,The Shawshank Redemption,1994,9.3
1,The Godfather,1972,9.2
2,The Dark Knight,2008,9.0
3,The Godfather: Part II,1974,9.0


Section 5: GroupBy & Aggregation (Q19–23)

19. Group by Year and find the average rating per year.

In [25]:
df.groupby('Released_Year')['IMDB_Rating'].mean()

Released_Year
1920    8.100000
1921    8.300000
1922    7.900000
1924    8.200000
1925    8.100000
          ...   
2017    7.890909
2018    7.994737
2019    7.995652
2020    8.133333
PG      7.600000
Name: IMDB_Rating, Length: 100, dtype: float64

20. Group by Director and find who has the most movies in the dataset.

In [26]:
df.groupby('Director')['Series_Title'].max()

Director
Aamir Khan                       Taare Zameen Par
Aaron Sorkin           The Trial of the Chicago 7
Abdellatif Kechiche                La vie d'Adèle
Abhishek Chaubey                      Udta Punjab
Abhishek Kapoor                       Kai po che!
                                  ...            
Zack Snyder                              Watchmen
Zaza Urushadze                        Mandariinid
Zoya Akhtar              Zindagi Na Milegi Dobara
Çagan Irmak                        Babam ve Oglum
Ömer Faruk Sorak                         G.O.R.A.
Name: Series_Title, Length: 548, dtype: object

21. Group by Genre and find the average revenue per genre.

In [27]:
df_exploded = df.assign(Genre=df['Genre'].str.split(', ')).explode('Genre')
df_exploded.groupby('Genre')['Gross'].mean().round()

Genre
Action       120687024.0
Adventure    146867962.0
Animation    108863266.0
Biography     51472288.0
Comedy        56807850.0
Crime         37562095.0
Drama         42107798.0
Family        83018119.0
Fantasy       95725031.0
Film-Noir     12728418.0
History       48108955.0
Horror        34855026.0
Music         32383187.0
Musical       20787249.0
Mystery       38585975.0
Romance       37935622.0
Sci-Fi       136877546.0
Sport         53672020.0
Thriller      51750095.0
War           34084890.0
Western       47521249.0
Name: Gross, dtype: float64

22. Find the year with the highest average movie rating.

In [28]:
df.head()

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,28341469.0
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,134966411.0
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,534858444.0
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,57300000.0
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,4360000.0


In [29]:
df.groupby('Released_Year')['IMDB_Rating'].mean().idxmax()

'1936'

23. Group by Director and find the average rating per director, sorted descending.

In [30]:
df.groupby('Director')['IMDB_Rating'].mean().sort_values(ascending=False)

Director
Frank Darabont     8.95
Irvin Kershner     8.70
Lana Wachowski     8.70
George Lucas       8.60
Roberto Benigni    8.60
                   ... 
Eric Bress         7.60
Ted Demme          7.60
Neil Burger        7.60
Bennett Miller     7.60
Mary Harron        7.60
Name: IMDB_Rating, Length: 548, dtype: float64

Section 6: Advanced/Combined (Q24–25)

24. Find the director with the highest average revenue (only consider directors with 3+ movies).

In [31]:
dirt_stats = df.groupby('Director')['Gross'].agg((['mean','count']))
dirt_stats = dirt_stats[dirt_stats['count']>=3]
dirt_stats['mean'].idxmax()

'Anthony Russo'

25. Create a new column Profitability (if Budget column exists) or Decade (from Year, e.g., 2010s, 2000s) and group movies by it to compare average ratings.

In [32]:
df[df['Released_Year'].str.isnumeric() == False]

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
966,https://m.media-amazon.com/images/M/MV5BNjEzYj...,Apollo 13,PG,U,140,"Adventure, Drama, History",7.6,NASA must devise a strategy to return Apollo 1...,77.0,Ron Howard,Tom Hanks,Bill Paxton,Kevin Bacon,Gary Sinise,269197,173837933.0


In [33]:
df = df[df['Released_Year'].str.isnumeric() == True]
df['Released_Year'] = df['Released_Year'].astype(int)

In [34]:
df['Decade'] = (df['Released_Year'] // 10) * 10
df.groupby('Decade')['IMDB_Rating'].mean().sort_values(ascending=False)

Decade
2020    8.133333
1920    8.127273
1950    8.058929
1940    8.025714
1960    7.973973
1970    7.969737
1930    7.966667
1990    7.963333
1980    7.953933
2010    7.924380
2000    7.899578
Name: IMDB_Rating, dtype: float64

In [35]:
df.columns

Index(['Poster_Link', 'Series_Title', 'Released_Year', 'Certificate',
       'Runtime', 'Genre', 'IMDB_Rating', 'Overview', 'Meta_score', 'Director',
       'Star1', 'Star2', 'Star3', 'Star4', 'No_of_Votes', 'Gross', 'Decade'],
      dtype='object')